wicketd/src/update_tracker.rs

```diff
@@ -346,6 +347,10 @@ impl UpdateDriver {
         self.update_log.lock().unwrap().current = Some(state);
     }
 
+    fn clear_current_update_state(&self) {
+        self.update_log.lock().unwrap().current = None;
```

Comment:
Hmm. I think wicket is currently assuming `current = None` means no update is running, but that's not true here since it could observe `current = None` in between installinator completing and the host phase 1 being installed. I think we'd be okay to remove this and the call to it altogether - as soon as the installinator channel is closed, we move into `install_host_phase_1_and_boot` which should set a new state.

Reply:
Oh interesting, I didn't realize that. My specific concern here is that it's possible there are long-ish periods where the installinator doesn't return a progress update. This would result in step completion being ahead of progress, which feels wonky to me and might show up as an inconsistent state in the wicket UI.

Maybe we should just document this for now.

Reply:
> Oh interesting, I didn't realize that. My specific concern here is that it's possible there are long-ish periods where the installinator returns completion but not progress update (e.g. if it hasn't reached out to peers). This would result in step completion being ahead of progress, which feels wonky to me and might show up as an inconsistent state in the wicket UI.

I don't think I follow - what does "returns completion but not progress update" mean?

In terms of step completion being ahead of progress, this is what the (admittedly extremely clunky) `push_update_success()` function is for - while the update log lock is held, it both pushes a completion event and sets the new state, so a caller can't observe a completion event associated with a stale progress message. The clunkyness comes from needing to know what the next state is going to be, though; e.g., https://github.com/oxidecomputer/omicron/blob/main/wicketd/src/update_tracker.rs#L395-L400 pushes an "update done" and sets the next state to "I'm going to reset", which is weird because it requires knowledge of both what comes before and what comes after.

Reply:
> I don't think I follow - what does "returns completion but not progress update" mean?

So let's say that the installinator has just finished downloading the host phase 2 image, and is moving to start downloading the control plane image. In that case, what's going to happen is:

1. The host phase 2 image fetcher sends a completion event to the progress tracker running inside the installinator.
2. As part of receiving the message, the progress tracker will record the completion event and clear progress.
3. Then, when the control plane image fetcher connects to the server, it sends a progress message to the progress tracker, which records the message.

If the installinator sends a report in between steps 2 and 3, it will send an empty progress message list. I haven't measured the delay in practice yet.

Some possible solutions to this are:

1. Measure the delay and don't bother addressing this at all. It's possible the race condition lasts just a few milliseconds and is unlikely to ever be hit in practice. (Though given that reports go out every second, even 10 millseconds a is 1% chance).
2. Change the update tracker data model to handle the case of a cleared progress message.
3. Build in the capability you mentioned in the update tracker: it needs to know what the next step is for every completion step.
4. In the installinator, always set a progress message. In other words, build the capability *in the installinator* to know what the next step is given a completion step.
5. (others that I haven't thought of yet)

I was proposing option 4 in my prior comment.

Reply:
Published an update which doesn't clear progress for now.